src/clr-angular/popover/dropdown/providers/dropdown-focus-handler.service.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { isPlatformBrowser } from '@angular/common';
+import { Inject, Injectable, Optional, PLATFORM_ID, Renderer2, SkipSelf } from '@angular/core';
+import { Observable, of, ReplaySubject } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { IfOpenService } from '../../../utils/conditional/if-open.service';
+import { customFocusableItemProvider } from '../../../utils/focus/focusable-item/custom-focusable-item-provider';
+import { UNIQUE_ID } from '../../../utils/id-generator/id-generator.service';
+import { ArrowKeyDirection } from '../../../utils/focus/arrow-key-direction.enum';
+import { FocusService } from '../../../utils/focus/focus.service';
+import { FocusableItem } from '../../../utils/focus/focusable-item/focusable-item';
+import { linkParent, linkVertical } from '../../../utils/focus/focusable-item/linkers';
+import { wrapObservable } from '../../../utils/focus/wrap-observable';
+
+@Injectable()
+export class DropdownFocusHandler implements FocusableItem {
+  constructor(
+    @Inject(UNIQUE_ID) public id: string,
+    private renderer: Renderer2,
+    @SkipSelf()
+    @Optional()
+    private parent: DropdownFocusHandler,
+    private ifOpenService: IfOpenService,
+    private focusService: FocusService,
+    @Inject(PLATFORM_ID) private platformId: Object
+  ) {
+    this.resetChildren();
+    this.moveToFirstItemWhenOpen();
+    if (!this.parent) {
+      this.handleRootFocus();
+    }
+  }
+
+  /**
+   * If the dropdown was opened by clicking on the trigger, we automatically move to the first item
+   */
+  moveToFirstItemWhenOpen() {
+    this.ifOpenService.openChange.subscribe(open => {
+      if (open && this.ifOpenService.originalEvent) {
+        // Even if we properly waited for ngAfterViewInit, the container still wouldn't be attached to the DOM.
+        // So setTimeout is the only way to wait for the container to be ready to move focus to first item.
+        setTimeout(() => {
+          this.focusService.moveTo(this);
```

Comment:
I see now two methods that use `setTimeout`. If possible, we should try to avoid it as much as possible. I understand why you needed them. But it seems like we could avoid both usages by placing these subscriptions in the container setter: 
```ts
set container(el: HTMLElement) {
  if(this.openChangeSubscription) {
       this.openChangeSubscription.unsubscribe();
  }
  this.openChangeSubscription = this.ifOpenService.openChange.subscribe(open => {  ...  });
  ....
}
```

Reply:
This doesn't work, because the openChange fires already once before the set container is called that we want to catch and act upon. I tried to make it execute the logic that the subscription would execute in this setter and then subscribe for the future events, but still some of the other states are inconsistent at that time of initial execution and produces error. For now, I'll create a separate ticket to see if we can clean up the setTimeout logic at a later time.

Reply:
New issue filed: https://github.com/vmware/clarity/issues/3362

Reply:
Sounds good. Thanks, Jeeyun 👌